go/interpreter/lfvm/memory_test.go

```diff
@@ -47,3 +52,232 @@ func TestExpansionCosts(t *testing.T) {
+		expected error
+	}{
+		"regular":        {value: 0x12, offset: 0, gas: 100, expected: nil},
+		"not enough gas": {value: 0x12, offset: 64, gas: 0, expected: errOutOfGas},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctxt := getEmptyContext()
+			m := NewMemory()
+			ctxt.gas = test.gas
+
+			err := m.SetByte(test.offset, test.value, &ctxt)
+			if !errors.Is(err, test.expected) {
+				t.Errorf("unexpected error, want: %v, got: %v", test.expected, err)
+			}
+			if err == nil && m.store[test.offset] != test.value {
+				t.Errorf("unexpected value, want: %v, got: %v", test.value, m.store[test.offset])
+			}
+		})
+	}
+}
+
+func TestSetWordProperlyReportsNotEnoughGas(t *testing.T) {
+	ctxt := getEmptyContext()
+	m := NewMemory()
+	ctxt.gas = 0
+	err := m.SetWord(0, uint256.NewInt(0), &ctxt)
+	if !errors.Is(err, errOutOfGas) {
+		t.Errorf("unexpected error, want: %v, got: %v", errOutOfGas, err)
+	}
+}
+
+func TestMemory_Set(t *testing.T) {
+
+	tests := map[string]struct {
+		size     uint64
+		offset   uint64
+		expected error
+	}{
+		"regular":         {size: 8, offset: 0, expected: nil},
+		"size overflow":   {size: math.MaxUint64, offset: 1, expected: errGasUintOverflow},
+		"offset overflow": {size: 32, offset: math.MaxUint64, expected: errGasUintOverflow},
+		"not enough memory": {size: 32, offset: 32,
+			expected: errSetMemTooSmall(8, 32, 32)},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctxt := getEmptyContext()
+			m := NewMemory()
+			m.store = make([]byte, 8)
+			ctxt.gas = 100
+
+			data := []byte{0x12, 0x34, 0x56, 0x78, 0x90, 0xab, 0xcd, 0xef}
+			err := m.Set(test.offset, test.size, data)
+
+			if err == nil {
+				if test.expected != nil {
+					t.Errorf("expected error %v, got nil", test.expected)
+				}
+				if m.Len() != test.size+test.offset {
+					t.Errorf("unexpected memory size, want: %d, got: %d", test.size+test.offset, m.Len())
+				}
+				if !bytes.Equal(m.store[test.offset:], data) {
+					t.Errorf("unexpected memory value, want: %x, got: %x", data, m.store[test.offset:])
+				}
+			} else if !errors.Is(err, test.expected) {
+				t.Errorf("unexpected error, want: %v, got: %v", test.expected, err)
+			}
+
+		})
+	}
+}
+
+func TestMemory_CopyData(t *testing.T) {
+
+	baseStore := []byte{0x12, 0x34, 0x56, 0x78, 0x90, 0xab, 0xcd, 0xef}
+
+	tests := map[string]struct {
+		store    []byte
+		offset   uint64
+		target   []byte
+		expected []byte
+	}{
+		"regular": {store: baseStore, offset: 0, target: make([]byte, 8),
+			expected: baseStore},
+		"offset bigger than memory": {store: baseStore, offset: 9, target: make([]byte, 8),
+			expected: []byte{0, 0, 0, 0, 0, 0, 0, 0}},
+		"padding needed": {store: baseStore,
+			offset: 4, target: make([]byte, 8),
+			expected: []byte{0x90, 0xab, 0xcd, 0xef, 0, 0, 0, 0}},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			m := NewMemory()
+			m.store = test.store
+			target := make([]byte, 8)
```

Comment:
nit: by not initializing with zeros, you can check that the memory is actually copied and not just expanded and copy suffers from some bonduary error.